### ISSUE TITLE
Interface Selection Not Working

### ISSUE THREAD

**Issue by jwbensley** [jwbensley/Etherate#6]
Via email:

...trying to use it to test 40Gb network connectivity but can’t get it to work:

```
[root@r6181-d5-us01 Etherate-master]# ./etherate

Running in TX mode

Error: Couldn't find appropriate interface ID, returned ID was 0!

 Try supplying a source MAC address with the -s option.

[root@r6181-d5-us01 Etherate-master]# ./etherate -s e4:1d:2d:af:c5:80

Running in TX mode

Error: Couldn't find appropriate interface ID, returned ID was 0!

 Try supplying a source MAC address with the -s option.

[root@r6181-d5-us01 Etherate-master]# uname -a

Linux r6181-d5-us01 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jwbensley** [comment]
I have requested the user provide the outputs from "ip l" and "etherate -l".

Response from user:

[root@r6181-d5-us01 tmp]# ip l

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT

    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00

2: em1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT qlen 1000

    link/ether 20:47:47:80:80:50 brd ff:ff:ff:ff:ff:ff

3: em2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT qlen 1000

    link/ether 20:47:47:80:80:51 brd ff:ff:ff:ff:ff:ff

4: em3: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT qlen 1000

    link/ether 20:47:47:80:80:52 brd ff:ff:ff:ff:ff:ff

5: em4: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT qlen 1000

    link/ether 20:47:47:80:80:53 brd ff:ff:ff:ff:ff:ff

6: p7p1: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master bond0 state UP mode DEFAULT qlen 1000

    link/ether 00:10:18:ed:24:30 brd ff:ff:ff:ff:ff:ff

7: p7p2: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master bond0 state UP mode DEFAULT qlen 1000

    link/ether 00:10:18:ed:24:30 brd ff:ff:ff:ff:ff:ff

8: p5p1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT qlen 1000

    link/ether e4:1d:2d:af:c5:80 brd ff:ff:ff:ff:ff:ff

9: p5p2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT qlen 1000

    link/ether e4:1d:2d:af:c5:81 brd ff:ff:ff:ff:ff:ff

10: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT

    link/ether 00:10:18:ed:24:30 brd ff:ff:ff:ff:ff:ff


[root@r6181-d5-us01 Etherate-master]# ./etherate -l

Device lo with address 00:00:00:00:00:00, interface index 1

Device em1 with address 00:00:47:80:80:50, interface index 2

Device em2 with address 00:00:47:80:80:51, interface index 3

Device em3 with address 00:00:47:80:80:52, interface index 4

Device em4 with address 00:00:47:80:80:53, interface index 5

Device p7p1 with address 00:00:18:ed:24:30, interface index 6

Device p7p2 with address 00:00:18:ed:24:30, interface index 7

Device p5p1 with address 00:00:2d:af:c5:80, interface index 8

Device p5p2 with address 00:00:2d:af:c5:81, interface index 9

Device bond0 with address 00:00:18:ed:24:30, interface index 10

**Comment by jwbensley** [comment]
The user was trying to use this interface:

```
8: p5p1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT qlen 1000

    link/ether e4:1d:2d:af:c5:80 brd ff:ff:ff:ff:ff:ff
```

The Etherate output shows the first two octets as zero:

`Device p5p1 with address 00:00:2d:af:c5:80, interface index 8`

The user has confirmed they can use the interface if they use the Etherate "i" CLI switch to specify the interface index numerically.

**Comment by tanev** [comment]
I guess it is better to report a similar issue here:

```
# ./etherate -l
# echo $?
0
```

```
# ip l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 0c:c4:7a:00:d8:5e brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 0c:c4:7a:00:d8:5f brd ff:ff:ff:ff:ff:ff
4: eth2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq portid f452140300769bb0 state UP mode DEFAULT group default qlen 1000
    link/ether f4:52:14:76:9b:b0 brd ff:ff:ff:ff:ff:ff
5: eth3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq portid f452140300769bb1 state UP mode DEFAULT group default qlen 1000
    link/ether f4:52:14:76:9b:b1 brd ff:ff:ff:ff:ff:ff
6: eth2.304@eth2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether f4:52:14:76:9b:b0 brd ff:ff:ff:ff:ff:ff
7: eth3.404@eth3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether f4:52:14:76:9b:b1 brd ff:ff:ff:ff:ff:ff
# ./etherate -i eth2.304 -s F4:52:14:76:9B:B0 -d 00:02:C9:3C:E3:80
Running in TX mode
Error: Couldn't set interface index from name, returned index was 0!
```

Another instance was running on a separate server with a simple `./etherate -r`. Note that without specifying the interface it were using eth0 by default.

Turned out that I could potentially check the index by testing with different values for `-I` and found that the `eth2.304` interface (which is a tagged VLAN interface) is at index 7 on one of the nodes and at index 6 at the other. However an attempt to test between these two nodes on this interface would not finish staying at:

- node1:

```
# ./etherate -r -i eth2.304 -I 7
Running in RX mode
Using device eth2.304 with address 00:00:c9:3c:e3:80, interface index 7
Entering promiscuous mode
Source MAC 00:00:5e:00:00:02
Destination MAC 00:00:5e:00:00:01
Sending gratuitous broadcasts...

Waiting for settings from TX host
Settings have been synchronised

Calculating delay between hosts...
```

- node2:

```
# ./etherate -i eth2.304 -I 6
Running in TX mode
Using device eth2.304 with address 00:00:14:76:9b:b0, interface index 6
Entering promiscuous mode
Source MAC 00:00:5e:00:00:01
Destination MAC 00:00:5e:00:00:02
Sending gratuitous broadcasts...

Synchronising settings with RX host
Settings have been synchronised

Calculating delay between hosts...
```

This is with the latest version on Ubuntu 16.04 LTS. Please let me know in case you need more information on any of the above.

Best regards,
Todor Tanev

**Comment by jwbensley** [comment]
Thanks for the info Todor and sorry about the delay, my day job has been very busy lately.

> Turned out that I could potentially check the index by testing with different values for -I and found that the eth2.304 interface (which is a tagged VLAN interface) is at index 7 on one of the nodes and at index 6 at the other. 

If you use the `ip -l` command on each host it should show you the different interface indexes on each host. There isn't anything I can do about them being different on each host, its up to the OS to assign the interface index.

For this issue:
```
# ./etherate -l
# echo $?
0
```

Please can you grab the latest version of Etherate and see if it you can see your interfaces now? I expect you should be able to with the correct interface index numbers although the MAC addresses might be wrong, the same as with the previous user above. It seems a change in the kernel behavior in newer kernels is causing the MACs to display incorrectly (I'm still looking into it). If you can grab the latest code and see if at least the interface indexes are correct shown that would be helpful, then at least you can run tests using the interface index.


> However an attempt to test between these two nodes on this interface would not finish staying at:
```
node1:
# ./etherate -r -i eth2.304 -I 7
```

Etherate is meant to write directly to the wire. Please can you instead use `./etherate -I 4 -v 304` on both sides to select the physical interface (eth2) and not the logical interface (eth2.304) and tell Etherate to add VLAN tag 304 to the frame headers?

**Comment by jwbensley** [comment]
Note to self:

`ioctl()` to grab the if index (`SIOCGIFINDEX`) when the ifr_name is already set seems to be the culprit. A quick initial test shows this is working, copy the MAC then run `ioctl()` to get the if index:

```
                char mac[6];
                memcpy(mac, ifreq.ifr_addr.sa_data, 6);

                // Get the interface index
                ioctl(SOCKET_FD, SIOCGIFINDEX, &ifreq);

                // Print the current_cpu_setnt interface details
                printf("Device %s with address %02x:%02x:%02x:%02x:%02x:%02x, "
                       "has interface index %d\n",
                       ifreq.ifr_name,
                       (uint8_t)mac[0],
                       (uint8_t)mac[1],
                       (uint8_t)mac[2],
                       (uint8_t)mac[3],
                       (uint8_t)mac[4],
                       (uint8_t)mac[5],
                       ifreq.ifr_ifindex);

```

**Comment by jwbensley** [comment]
Replicated and fixed on:

```
me@lsrv-ld5-mon-01:~/testing$ uname -a
Linux lsrv-ld5-mon-01 3.5.0-37-generic #58~precise1-Ubuntu SMP Wed Jul 10 17:48:11 UTC 2013 x86_64 x86_64 x86_64 GNU/Linux

me@lsrv-ld5-mon-01:~/testing$ cat /etc/issue
Ubuntu 12.04.2 LTS \n \l
```

**Comment by jwbensley** [comment]
Fixed in beta 0.11.
